### PR TITLE
overlord/snapstate: validate instance names early

### PR DESCRIPTION
**Pull request by bboozzoo**
We already validate instance name when installing snap from a path, but the
store installation paths were missing similar treatment. As a result, snap
installation would fail very late in the process, reaching the mount task.

The patch introduces makes to validate the snap names early for store
installation code paths.

### CODE REVIEW

File: overlord/snapstate/snapstate.go [hunk 1]
```diff
@@ -622,6 +622,10 @@ func Install(st *state.State, name, channel string, revision snap.Revision, user
 		return nil, err
 	}
 
+	if err := snap.ValidateInstanceName(name); err != nil {
+		return nil, err
```
Comment (stolowski) [inline overlord/snapstate/snapstate.go]: Should the error here be decorated similiar to the case below ("fmt.Errorf("invalid instance name:....."))?

Comment (bboozzoo) [reply overlord/snapstate/snapstate.go]: It's a single snap install, so in contrast to `InstallMany()`, the error implicitly is about the current snap. That was my reasoning, but I'm happy to tweak it as needed.

File: overlord/snapstate/snapstate.go [hunk 2]
```diff
@@ -660,6 +664,11 @@ func InstallMany(st *state.State, names []string, userID int) ([]string, []*stat
 		if snapst.IsInstalled() {
 			continue
 		}
+
+		if err := snap.ValidateInstanceName(name); err != nil {
+			return nil, nil, fmt.Errorf("invalid instance name: %q", name)
```
Comment (stolowski) [inline overlord/snapstate/snapstate.go]: I think we should include the original err here.

File: overlord/snapstate/snapstate_test.go [hunk 3]
```diff
@@ -12944,6 +12944,31 @@ func (s *snapmgrTestSuite) TestSnapManagerCanStandby(c *C) {
...
+	defer s.state.Unlock()
+
+	_, err := snapstate.Install(s.state, "foo--invalid", "", snap.R(0), 0, snapstate.Flags{})
+	c.Assert(err, ErrorMatches, `invalid snap name: "foo--invalid"`)
+
+	_, err = snapstate.Install(s.state, "foo_123_456", "", snap.R(0), 0, snapstate.Flags{})
+	c.Assert(err, ErrorMatches, `invalid instance key: "123_456"`)
+
+	_, _, err = snapstate.InstallMany(s.state, []string{"foo--invalid"}, 0)
+	c.Assert(err, ErrorMatches, `invalid instance name: "foo--invalid"`)
```
Comment (stolowski) [inline overlord/snapstate/snapstate_test.go]: Once the error propagation is fixed, both Install and InstallMany should give same error messages with same inputs.

**Comment by pedronis** [comment]
@zyga 
> Should installation of invalid snap name be an early, synchronous error?

all the places here are early, they are not in task handlers

**Comment by bboozzoo** [comment]
I'll open a separate PR with an updated spread tests.

**Pull request merged**